# Teste 2 - 11/09/2018

## Disponibilização: 10/09/2018 - 11h
## Encerramento: 11/09/2018 - 20h

O objetivo deste segundo projeto prático da disciplina Redes Neurais Artificias é praticar os conceitos de Machine Learning vistos até o momento, em especial aqueles relativos ao processo de Aprendizagem de Máquina.

Vamos trabalhar com o dataset **Breast Cancer Wisconsin (Diagnostic) Data Set**, vide: <a href="https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)">Repositório UCI</a>

Esta tarefa é dividida em to-dos, isto é, pequenas atividades que devem ser cumpridas para que o objetivo geral seja alcançado. A cada to-do está associada uma célula do Jupyter Notebook, que deve ser preenchida com código Python atendendo ao que se pede.


Edite aqui o nome da equipe:

- Gabriel Faraco
- Juliany Raiol

In [1]:
# Organize seus imports nesta célula
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier, NearestCentroid
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import confusion_matrix

### To-Do 1

1. Você deve importar o dataset a partir do sci-kit learn.
Consulte o link: [Link da documentação do sci-kit learn](http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer)
   * Este dataset está organizado sob a forma de um dicionário, em que os dados preditores encontram-se na chave 'data', composta de diversas matrizes. Cada matriz está associada a um nome 'feature_names'. 
2. Crie um novo dicionário que mapeia cada 'feature_name' para uma matriz correspondente.
    * Antes de fazer esta associação, transponha a matriz localizada na chave 'data' para obter a dimensão correta.
3. Transforme o dataset em um objetivo tipo DataFrame do pandas
4. Adicione o atributo-alvo ao dataset existente.
    * Importante: O atributo-alvo está na chave 'target' do dicionário, com nome 'target_names'

In [2]:
cancer = load_breast_cancer()

In [3]:
dict_cancer = {}
data = np.array(cancer.data).T

index = 0
for feature_names in cancer.feature_names:
    dict_cancer[feature_names] = data[index]
    index+=1

In [4]:
dataframe = pd.DataFrame(data = dict_cancer)

In [5]:
target_names = cancer.target
dataframe['target_names'] = target_names

In [6]:
dataframe.rename(columns=lambda x : x.replace(" ", "_"), inplace=True)

In [7]:
dataframe.head()

mean_radius  mean_texture  mean_perimeter  mean_area  mean_smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean_compactness  mean_concavity  mean_concave_points  mean_symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean_fractal_dimension      ...       worst_texture  worst_perimeter  \
0                 0.07871      ...               17.33           184.60   
1                 0.05667      ...               23.41           158.80   
2                 0.05999      ...               25.53           152.50   
3                 0.09744      ...               26.50            98.87   
4                 0.05883      ...               16.67           152.20   

   worst_area  worst_smoothness  worst_compactness  worst_concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst_concave_points  worst_symmetry  worst_fractal_dimension  target_names  
0                0.2654          0.4601                  0.11890             0  
1                0.1860          0.2750                  0.08902             0  
2                0.2430          0.3613                  0.08758             0  
3                0.2575          0.6638                  0.17300             0  
4                0.1625          0.2364                  0.07678             0  

[5 rows x 31 columns]

### To-Do 2

Utilizando `pandas.DataFrame` para manipular o dataset, faça o que se pede:
1. Informe a quantidade de exemplos existentes no dataset
2. Enumere os atributos existentes no dataset
3. Identifique o atributo-alvo e imprima-o
4. O dataset é balanceado?
5. Remova todos os atributos que contenham a palavra `error`

In [8]:
size = dataframe.shape[0]
print("Há %d exemplos no dataset" %size)

Há 569 exemplos no dataset


In [9]:
attributes = dataframe.shape[1]
print("O dataset possui %d atributos" %attributes)

O dataset possui 31 atributos


In [10]:
print('O atributo alvo é *target_names*')
dataframe.target_names

O atributo alvo é *target_names*


0      0
1      0
2      0
3      0
4      0
5      0
6      0
7      0
8      0
9      0
10     0
11     0
12     0
13     0
14     0
15     0
16     0
17     0
18     0
19     1
20     1
21     1
22     0
23     0
24     0
25     0
26     0
27     0
28     0
29     0
      ..
539    1
540    1
541    1
542    1
543    1
544    1
545    1
546    1
547    1
548    1
549    1
550    1
551    1
552    1
553    1
554    1
555    1
556    1
557    1
558    1
559    1
560    1
561    1
562    0
563    0
564    0
565    0
566    0
567    0
568    1
Name: target_names, Length: 569, dtype: int64

In [11]:
allDiagnostics = dataframe.shape[0]

benign = dataframe.loc[dataframe.target_names == 0] 
benign = len(benign)

malignant = allDiagnostics - benign
benign, malignant

(212, 357)

In [12]:
print("Percentual de Diagnóstico Negativo de Câncer (0): %.2f" %(benign*100/allDiagnostics))
print("Percentual de Diagnóstico Positivo de Câncer (1): %.2f" %(malignant*100/allDiagnostics))

Percentual de Diagnóstico Negativo de Câncer (0): 37.26
Percentual de Diagnóstico Positivo de Câncer (1): 62.74


Com base nos dados apresentados anteriormente, observa-se que o dataset não é balanceado, pois o número de valores negativos (não dignosticados com câncer) é menor que os positivos (dignosticados com câncer). Esses últimos representam 62.74% dos dados contidos na base de dados. Para ter balanceamento, é recomendado 50% para cada label do atributo alvo.

In [38]:
erroList = list(dataframe.filter(regex='error'))
dataframe.drop(erroList, inplace=True, axis=1)
dataframe.head()

mean_radius  mean_texture  mean_perimeter  mean_area  mean_smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean_compactness  mean_concavity  mean_concave_points  mean_symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean_fractal_dimension  worst_radius  worst_texture  worst_perimeter  \
0                 0.07871         25.38          17.33           184.60   
1                 0.05667         24.99          23.41           158.80   
2                 0.05999         23.57          25.53           152.50   
3                 0.09744         14.91          26.50            98.87   
4                 0.05883         22.54          16.67           152.20   

   worst_area  worst_smoothness  worst_compactness  worst_concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst_concave_points  worst_symmetry  worst_fractal_dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678

### To-Do 3

Faça uma partição randomizada do tipo 70/30 para conjunto de treinamento e de testes.
Em ambos os conjuntos, separe o atributo-alvo.

Para facilitar, siga a nomenclatura sugerida:
* X_train: atributos preditores para o conjunto de treinamento
* X_test: atributos preditores para o conjunto de testes
* Y_train: atributo-alvo para os exemplos do conjunto de treinamento
* Y_test: atributo-alvo para os exemplos do conjunto de testes

Sugestão: [consultar a documentação do sci-kit learn](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#sklearn.model_selection.train_test_split)


In [35]:
Y = dataframe.target_names
dataframe.drop(['target_names'], axis=1, inplace=True)

X = dataframe
X.head()

AttributeError: 'DataFrame' object has no attribute 'target_names'

In [36]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape, Y_train.shape, Y_test.shape

((455, 20), (114, 20), (455,), (114,))

### To-Do 4

Vamos usar os dados X_train e Y_train para treinar dois modelos diferentes de Aprendizagem de Máquina.
1. Modelo 1: Vizinhos mais próximos, com k = 5
2. Modelo 2: Centróides mais próximos, de acordo com a distância Euclidiana

Basta completar o código a seguir

In [16]:
# 5 - vizinhos mais próximos
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, Y_train)

# Kernel Density
nc = NearestCentroid()
nc.fit(X_train, Y_train)

NearestCentroid(metric='euclidean', shrink_threshold=None)

### To-Do 5

Utilizar o conjunto de testes para prever o conjunto de testes

In [17]:
previsaokNN = knn.predict(X_test)
previsaonc = nc.predict(X_test)

### To-Do 6

Analisando as diferenças e igualdades entre os vetores previsaokNN, previsaonc e Y_test, construa as matrizes de confusão dos respectivos modelos de Machine Learning. 

Consulte: [Documentação do sklearn para Matrizes de Confusão](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)

In [18]:
confusion_matrix(Y_test, previsaokNN, labels=[0,1])

array([[ 58,   5],
       [  2, 106]])

In [19]:
confusion_matrix(Y_test, previsaonc,  labels=[0,1])

array([[ 49,  14],
       [  0, 108]])

### To-Do 7

Para cada um dos modelos, apresente:

1. Acurácia
2. Precisão
3. Revocação
4. F-Score (Leve em consideração se o dataset é balanceado ou não)

Consulte: [Documentação do sklearn para Métricas de Desempenho](http://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics)

In [20]:
print("Métricas do Modelo KNN")
print("acuracia: ", accuracy_score(Y_test, previsaokNN))
print("precisão: ", precision_score(Y_test, previsaokNN))
print("revocação: ", recall_score(Y_test, previsaokNN))
print("F-Score: ", f1_score(Y_test, previsaokNN, average='micro'))

print("\nMétricas do NC")
print("acuracia: ", accuracy_score(Y_test, previsaonc))
print("precisão: ", precision_score(Y_test, previsaonc))
print("revocação: ", recall_score(Y_test, previsaonc))
print("F-Score: ", f1_score(Y_test, previsaonc, average='micro'))

Métricas do Modelo KNN
acuracia:  0.9590643274853801
precisão:  0.954954954954955
revocação:  0.9814814814814815
F-Score:  0.9590643274853801

Métricas do NC
acuracia:  0.9181286549707602
precisão:  0.8852459016393442
revocação:  1.0
F-Score:  0.9181286549707602


### To-Do 8

Utilizando argumentos textuais, justifique qual dos modelos apresentados é melhor para o problema em questão.

Apesar de o KNN apresentar melhores valores que o CN com relação à acurácia, precisão e F-score, este também possui maior erro ao identificar <strong>casos malignos de câncer </strong>. De acordo com a matriz de confusão apresentada anteriormente, o CN não possuiu *falsos positivos*. Considerando o cenário, conclui-se que o modelo CN é o mais recomentado para o problema em questão.